xen/arch/x86/slaunch.c

```diff
@@ -0,0 +1,276 @@
+#include <xen/types.h>
+#include <asm/e820.h>
+#include <asm/intel_txt.h>
+#include <asm/page.h>
+#include <asm/processor.h>
+#include <asm/slaunch.h>
+#include <xen/init.h>
+#include <xen/mm.h>
+#include <xen/multiboot.h>
+
+/* SLB is 64k, 64k-aligned */
+#define SKINIT_SLB_SIZE  0x10000
+#define SKINIT_SLB_ALIGN 0x10000
+
+bool __initdata slaunch_active;
+uint32_t __initdata slaunch_param;
+
+static void __maybe_unused compile_time_checks(void)
+{
+    BUILD_BUG_ON(sizeof(slaunch_active) != 1);
+}
+
+int __init map_l2(unsigned long paddr, unsigned long size)
+{
+    unsigned long aligned_paddr = paddr & ~((1ULL << L2_PAGETABLE_SHIFT) - 1);
+    unsigned long pages = ((paddr + size) - aligned_paddr);
+    pages = ROUNDUP(pages, 1ULL << L2_PAGETABLE_SHIFT) >> PAGE_SHIFT;
+
+    if ( (aligned_paddr + pages * PAGE_SIZE) <= PREBUILT_MAP_LIMIT )
+        return 0;
+
+    if ( aligned_paddr < PREBUILT_MAP_LIMIT )
+    {
+        pages -= (PREBUILT_MAP_LIMIT - aligned_paddr) >> PAGE_SHIFT;
+        aligned_paddr = PREBUILT_MAP_LIMIT;
+    }
+
+    return map_pages_to_xen((unsigned long)__va(aligned_paddr),
+                            maddr_to_mfn(aligned_paddr),
+                            pages, PAGE_HYPERVISOR);
+}
+
+static uint32_t get_slb_start(void)
+{
+    /* The runtime computation relies on size being a power of 2 and equal to
+     * alignment. Make sure these assumptions hold. */
+    BUILD_BUG_ON(SKINIT_SLB_SIZE != SKINIT_SLB_ALIGN);
+    BUILD_BUG_ON(SKINIT_SLB_SIZE == 0);
+    BUILD_BUG_ON((SKINIT_SLB_SIZE & (SKINIT_SLB_SIZE - 1)) != 0);
+
+    /* Rounding any address within SLB down to alignment gives SLB base and
+     * slaunch_param points to SLRT inside SLB on AMD. */
+    return slaunch_param & (SKINIT_SLB_SIZE - 1);
+}
+
+void __init map_skinit_mem_regions(void)
```

Comment:
SLB is not enough, Xen needs to have at least TPM register space (0xfed40000-0xfed44fff) mapped:

```
(XEN) Early fatal page fault at e008:ffff82d0403133a1 (cr2=ffff8300fed40014, ec=0000)
(XEN) ----[ Xen-4.17.3  x86_64  debug=n  Not tainted ]----
(XEN) CPU:    0
(XEN) RIP:    e008:[<ffff82d0403133a1>] tpm_hash_extend+0x321/0x7f0
(XEN) RFLAGS: 0000000000010086   CONTEXT: hypervisor
(XEN) rax: ffff8300fed40014   rbx: ffff830003284000   rcx: 00000000fed40014
(XEN) rdx: 0000000000000000   rsi: 0000000000000002   rdi: ffff830000000000
(XEN) rbp: ffff82d0403e7d80   rsp: ffff82d0403e7b78   r8:  ffffffffffffffff
(XEN) r9:  ffff83000011e5bc   r10: 0000000000000000   r11: 0000000000008000
(XEN) r12: 0000000000000502   r13: ffff83000011e5bc   r14: 00000000009aa4a0
(XEN) r15: 0000000000400000   cr0: 0000000080050033   cr4: 00000000000000a0
(XEN) cr3: 00000000dbc43000   cr2: ffff8300fed40014
(XEN) fsb: 0000000000000000   gsb: 0000000000000000   gss: 0000000000000000
(XEN) ds: 0000   es: 0000   fs: 0000   gs: 0000   ss: 0000   cs: e008
(XEN) Xen code around <ffff82d0403133a1> (tpm_hash_extend+0x321/0x7f0):
(XEN)  d4 fe 48 09 c8 48 01 f8 <8b> 00 a9 00 00 00 50 0f 84 34 fe ff ff e9 f9 fd
```

In fact, it is also required on TXT, but because TXT register space is located in the same 2MB page as TPM, it went unnoticed.

Reply:
165c86c12102b283244173b63d169b20322fe534 and https://github.com/TrenchBoot/xen/compare/4de56cbbf531cf9efb2efd4baf58f66db71b5ae9..6c44eb9bcc4f73b6b98e6a6cfd50c0994c85580c#diff-1d266d1dff9daf02e019100b9c620a099561c98811a05a0b7581363b3c00ac39R59